src/full/Agda/Compiler/Malfunction/Compiler.hs

```diff
@@ -0,0 +1,627 @@
+--
+-- Example
+-- λλ Env{_level = 2} usedVars (λ(λ ((Var 3) (λ (Var 4)))) ) == {1}
+usedVars :: MonadReader Env m => TTerm -> m (Set Int)
+usedVars term = asks _level >>= go mempty term
+   where
+     go vars t topnext = goterm vars t
+       where
+         goterms vars = foldM (\acvars tt -> goterm acvars tt) vars
+         goterm vars t = do
+           nextix <- asks _level
+           case t of
+             (TVar v) -> return $ govar vars v nextix
+             (TApp t args) -> goterms vars (t:args)
+             (TLam t) -> snd <$> introVar (goterm vars t)
+             (TLet t1 t2) -> do
+               vars1 <- goterm vars t1
+               snd <$> introVar (goterm vars1 t2)
+             (TCase v _ def alts) -> do
+               vars1 <- goterm (govar vars v nextix) def
+               foldM (\acvars alt -> goalt acvars alt) vars1 alts
+             _ -> return vars
+         govar vars v nextix
+           | 0 <= v' && v' < topnext = Set.insert v' vars
+           | otherwise = vars
+           where v' = v + topnext - nextix
+         goalt vars alt = case alt of
+           TACon _ _ b -> goterm vars b
+           TAGuard g b -> goterms vars [g, b]
+           TALit{} -> return vars
+
+
+-- TODO: Translate constructors differently from names.
+-- Don't know if we should do the same when translating TDef's, but here we
+-- should most likely use malfunction "blocks" to represent constructors
+-- in an "untyped but injective way". That is, we only care that each
+-- constructor maps to a unique number such that we will be able to
+-- distinguish it in malfunction. This also means that we should carry
+-- some state around mapping each constructor to it's corresponding
+-- "block-representation".
+--
+-- An example for clarity. Consider type:
+--
+--   T a b = L a | R b | B a b | E
+--
+-- We need to encode the constructors in an injective way and we need to
+-- encode the arity of the constructors as well.
+--
+--   translate (L a)   = (block (tag 2) (tag 0) a')
+--   translate (R b)   = (block (tag 2) (tag 1) b')
+--   translate (B a b) = (block (tag 3) (tag 2) a' b')
+--   translate E       = (block (tag 1) (tag 3))
+-- TODO: If the length of `ts` does not match the arity of `nm` then a lambda-expression must be returned.
+translateCon :: MonadReader Env m => QName -> [TTerm] -> m Term
+translateCon nm ts = do
+  builtinbool <- builtinBool (qnameNameId nm)
+  case builtinbool of
+    Just t -> return (boolT t)
+    Nothing -> do
+      ts' <- mapM translateTerm ts
+      tag <- askConTag nm
+      arity <- askArity nm
+      let diff = arity - length ts'
+          vs   = take diff $ map pure ['a'..]
+      return $ if diff == 0
+      then Mblock tag ts'
+      else Mlambda vs (Mblock tag (ts' ++ map Mvar vs))
+
+-- | Ugly hack to represent builtin bools as integers.
+-- For now it checks whether the concrete name ends with "Bool.true" or "Bool.false"
+builtinBool :: MonadReader Env m => NameId -> m (Maybe Bool)
```

Comment:
You can use the predefined builtins for this:
https://github.com/agda/agda/blob/c297fe095a76ddcd00766df6eeffdf958d272273/src/full/Agda/TypeChecking/Monad/Builtin.hs#L404
https://github.com/agda/agda/blob/master/src/full/Agda/Compiler/Treeless/EliminateLiteralPatterns.hs#L27